homeassistant/components/scheduler/schedule.py

```diff
@@ -0,0 +1,334 @@
+        # if we're still inside it; if we are, we're active:
+        last_end_dt = last_start_dt + self.instance_duration
+        return last_start_dt <= now <= last_end_dt
+
+    @property
+    def expired(self) -> bool:
+        """Return whether the schedule has expired."""
+        now = dt_util.utcnow()
+
+        # If a recurrence exists and there's at least one instance of it in the future,
+        # we're not expired:
+        if self.recurrence and self.recurrence.after(now, inc=True):
+            return False
+
+        # If either the initial start datetime or the end datetime (if it exists) are in
+        # in the future, we're not expired:
+        if self.start_datetime >= now or (
+            self.end_datetime and self.end_datetime >= now
+        ):
+            return False
+
+        return True
+
+    @callback
+    def _get_next_instance_datetimes(
+        self,
+    ) -> Tuple[Optional[datetime], Optional[datetime]]:
+        """Get the next starting (and, optionally, ending) datetimes."""
+        if not self._initial_instance_scheduled:
+            self._initial_instance_scheduled = True
+            return (self.start_datetime, self.end_datetime)
+
+        if self.recurrence:
+            start = self.recurrence.after(dt_util.utcnow(), inc=True)
+
+            # The recurrence has reached its end:
+            if not start:
+                return (None, None)
+
+            if self.end_datetime:
+                end = start + self.instance_duration
+            else:
+                end = None
+
+            return (start, end)
+
+        return (None, None)
+
+    def as_dict(self) -> dict:
+        """Return the schedule as a dict."""
+        the_dict = {
+            CONF_ENTITY_ID: self.entity_id,
+            CONF_START_DATETIME: self.start_datetime.isoformat(),
+        }
+
+        if self.end_datetime:
+            the_dict[CONF_END_DATETIME] = self.end_datetime.isoformat()
+        else:
+            the_dict[CONF_END_DATETIME] = None
+
+        if self.recurrence:
+            the_dict[CONF_RECURRENCE] = str(self.recurrence)
+        else:
+            the_dict[CONF_RECURRENCE] = None
+
+        return the_dict
+
+    @callback
+    def async_cancel(self) -> None:
+        """Cancel the schedule."""
+        self.active_instance.async_cancel()
+        self.active_instance = None
+
+        if self._async_revert_listener:
+            self._async_revert_listener()
+            self._async_revert_listener = None
+        if self._async_trigger_listener:
+            self._async_trigger_listener()
+            self._async_trigger_listener = None
+
+    @callback
+    def async_schedule(self) -> None:
+        """Schedule the next instance."""
+        start_dt, end_dt = self._get_next_instance_datetimes()
+
+        if not start_dt:
+            _LOGGER.info("No more instances of schedule: %s", self)
+            return
+
+        self.active_instance = ScheduleInstance(self._hass, self.entity_id)
```

Comment:
Should we still store the instance as `self.actige_instance` if there is no end time ?

Reply:
Good point!

Reply:
Actually, if we do that, we run into a case where `self.active_instance` will be `None`. In the case of `async_trigger`, because it will no longer be inline, it won't have access to that instance object.

Reply:
There are 2 time scheduling going on.

An ActiveInstance should be turned off if it has an end time.
An ActiveInstance needs to be triggered when its the start time.

We only need an active instance if a schedule has an end time. Otherwise we can just call the scene?